core/services/keystore/keys/ocr2key/generic_key_bundle.go

```diff
@@ -0,0 +1,176 @@
+package ocr2key
+
+import (
+	cryptorand "crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/services/keystore/chaintype"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2/types"
+)
+
+type (
+	keyring interface {
+		ocrtypes.OnchainKeyring
+		marshal() ([]byte, error)
+		unmarshal(in []byte) error
+	}
+
+	keyBundle[K keyring] struct {
+		keyBundleBase
+		keyring K
+	}
+
+	keyBundleRawData struct {
+		ChainType       chaintype.ChainType
+		OffchainKeyring []byte
+		Keyring         []byte
+
+		// old chain specific format for migrating
+		EVMKeyring    []byte `json:",omitempty"`
+		SolanaKeyring []byte `json:",omitempty"`
+		TerraKeyring  []byte `json:",omitempty"`
+	}
+)
+
+var _ KeyBundle = &keyBundle[*evmKeyring]{}
+var _ KeyBundle = &keyBundle[*solanaKeyring]{}
+var _ KeyBundle = &keyBundle[*terraKeyring]{}
+var _ KeyBundle = &keyBundle[*starknetKeyring]{}
+
+func newKeyBundle[K keyring](chain chaintype.ChainType, newKeyring func(material io.Reader) (K, error)) (*keyBundle[K], error) {
+	return newKeyBundleFrom(chain, newKeyring, cryptorand.Reader, cryptorand.Reader, cryptorand.Reader)
+}
+
+func mustNewKeyBundleInsecure[K keyring](chain chaintype.ChainType, newKeyring func(material io.Reader) (K, error), reader io.Reader) *keyBundle[K] {
+	key, err := newKeyBundleFrom(chain, newKeyring, reader, reader, reader)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to generate new OCR2-%s Key", chain))
+	}
+	return key
+}
+
+func newKeyBundleFrom[K keyring](chain chaintype.ChainType, newKeyring func(material io.Reader) (K, error), onchainSigningKeyMaterial, onchainEncryptionKeyMaterial, offchainKeyMaterial io.Reader) (*keyBundle[K], error) {
+	offchainKeyring, err := newOffchainKeyring(onchainSigningKeyMaterial, onchainEncryptionKeyMaterial)
+	if err != nil {
+		return nil, err
+	}
+	kr, err := newKeyring(onchainSigningKeyMaterial)
```

Comment:
this is the generic version of the previously mentioned key generation. if there are changes it should be done here